File: leetcode/array/easy/java_solutions/IntersectionOfTwoArrays_349.java
package leetcode.array.easy.java_solutions;

import java.util.*;

public class IntersectionOfTwoArrays_349 {

    static class Solution {
        public int[] intersection(int[] nums1, int[] nums2) {
            Set<Integer> seen = new HashSet<>();

            for (Integer n : nums1) {
                seen.add(n);
            }
            List<Integer> temp = new ArrayList<>();
            for (Integer n : nums2) {
                if (seen.contains(n)) {
                    temp.add(n);
                    seen.remove(n);
                }
            }
            int n = temp.size();
            int[] result = new int[n];
            for (int i = 0; i < n; i++) {
                result[i] = temp.get(i);
            }
            return result;
        }
    }
}

File: leetcode/array/easy/golang_solutions/intersection_of_two_arrays_349.go
package golang_solutions

func intersection(nums1 []int, nums2 []int) []int {
	m := make(map[int]interface{})
	for _, v := range nums1 {
		m[v] = nil
	}
	result := make([]int, 0)
	for _, v := range nums2 {
		if _, ok := m[v]; ok {
			result = append(result, v)
			delete(m, v)
		}
	}
	return result
}
